<launch>

  <arg name="gpu" default="0"/>
  <arg name="model_dir" default="$(find mirror_recognition)/logs/20190115-011612"/>
  <arg name="rviz" default="true"/>

  <!-- Republish images -->
  <group ns="kinect_head_remote">
    <node name="republish_rgb"
          pkg="image_transport" type="republish"
          args="compressed raw"
          output="screen"
          respawn="true">
      <remap from="in" to="/kinect_head/rgb/image_rect_color"/>
      <remap from="out" to="rgb/image_rect_color"/>
    </node>
    <node name="republish_depth"
          pkg="image_transport" type="republish"
          args="compressedDepth raw"
          output="screen"
          respawn="true">
      <remap from="in" to="/kinect_head/depth_registered/hw_registered/image_rect"/>
      <remap from="out" to="depth_registered/hw_registered/image_rect"/>
    </node>
  </group>

  <!-- Main process -->
  <include file="$(find mirror_recognition)/launch/include/fcn_mirror_segmentation_depth_estimation.launch">
    <arg name="INPUT_IMAGE" value="/kinect_head_remote/rgb/image_rect_color"/>
    <arg name="INPUT_DEPTH" value="/kinect_head_remote/depth_registered/hw_registered/image_rect"/>
    <arg name="model_dir" value="$(arg model_dir)"/>
    <arg name="gpu" value="$(arg gpu)"/>
  </include>

  <!-- Visualization -->
  <group if="$(arg rviz)">
    <node name="rviz"
          pkg="rviz" type="rviz"
          args="-d $(find mirror_recognition)/rvizconfig/mirror_segmentation_with_depth_by_pr2_kinect.rviz"/>

    <!-- Visualize input depth image -->
    <group ns="kinect_head_remote/depth_registered/hw_registered">
      <node name="image_rect_viz"
            pkg="image_view" type="image_view"
            output="screen"
            respawn="true">
        <remap from="image" to="image_rect"/>
        <rosparam>
          gui: false
          colormap: 2
        </rosparam>
      </node>
    </group>

    <!-- Visualize output depth image -->
    <group ns="fcn_mirror_segmentation_depth_estimation/output">
      <node name="depth_inpainted_viz"
            pkg="image_view" type="image_view"
            output="screen"
            respawn="true">
        <remap from="image" to="depth_inpainted"/>
        <rosparam>
          gui: false
          colormap: 2
        </rosparam>
      </node>
      <node name="depth_pred_raw_viz"
            pkg="image_view" type="image_view"
            output="screen"
            respawn="true">
        <remap from="image" to="depth_pred_raw"/>
        <rosparam>
          gui: false
          colormap: 2
        </rosparam>
      </node>
      <node name="depth_pred_labeled_viz"
            pkg="image_view" type="image_view"
            output="screen"
            respawn="true">
        <remap from="image" to="depth_pred_labeled"/>
        <rosparam>
          gui: false
          colormap: 2
        </rosparam>
      </node>
    </group>

    <!-- ROS console output -->
    <node name="rosconsole_overlay_text"
          pkg="jsk_rviz_plugins" type="rosconsole_overlay_text.py"
          output="screen"
          respawn="true">
      <rosparam>
        reverse_lines: false
      </rosparam>
    </node>
  </group>

</launch>
